File: domain/src/main/java/no/nav/melosys/domain/brev/VarselbrevManglendeInnbetalingBrevbestilling.java
package no.nav.melosys.domain.brev;

import java.time.LocalDate;

import no.nav.melosys.domain.manglendebetaling.Betalingsstatus;

public class VarselbrevManglendeInnbetalingBrevbestilling extends DokgenBrevbestilling {
    private String fakturanummer;
    private Betalingsstatus betalingsstatus;
    private String fullmektigForBetaling;
    private LocalDate betalingsfrist;

    public VarselbrevManglendeInnbetalingBrevbestilling() {
        super();
        //Tom constructor på grunn av deserialsering i prosessinstans
    }

    private VarselbrevManglendeInnbetalingBrevbestilling(VarselbrevManglendeInnbetalingBrevbestilling.Builder builder) {
        super(builder);
    }

    public String getFakturanummer() {
        return fakturanummer;
    }

    public Betalingsstatus getBetalingsstatus() {
        return betalingsstatus;
    }
    public String getFullmektigForBetaling() {
        return fullmektigForBetaling;
    }

    public LocalDate getBetalingsfrist() {
        return betalingsfrist;
    }

    public static final class Builder extends DokgenBrevbestilling.Builder<Builder> {
        private String fakturanummer;
        private Betalingsstatus betalingsstatus;
        private String fullmektigForBetaling;
        private LocalDate betalingsfrist;

        public Builder() {
        }

        public Builder(VarselbrevManglendeInnbetalingBrevbestilling varselbrevManglendeInnbetalingBrevbestilling) {
            super(varselbrevManglendeInnbetalingBrevbestilling);
            this.betalingsstatus = varselbrevManglendeInnbetalingBrevbestilling.getBetalingsstatus();
            this.fakturanummer = varselbrevManglendeInnbetalingBrevbestilling.getFakturanummer();
            this.fullmektigForBetaling = varselbrevManglendeInnbetalingBrevbestilling.getFullmektigForBetaling();
            this.betalingsfrist = varselbrevManglendeInnbetalingBrevbestilling.getBetalingsfrist();
        }

        public Builder medFakturanummer(String fakturanummer) {
            this.fakturanummer = fakturanummer;
            return this;
        }

        public Builder medBetalingsstatus(Betalingsstatus betalingsstatus) {
            this.betalingsstatus = betalingsstatus;
            return this;
        }

        public Builder medFullmektigForBetaling(String fullmektigForBetaling) {
            this.fullmektigForBetaling = fullmektigForBetaling;
            return this;
        }

        public Builder medBetalingsfrist(LocalDate betalingsfrist) {
            this.betalingsfrist = betalingsfrist;
            return this;
        }

        @Override
        public VarselbrevManglendeInnbetalingBrevbestilling build() {
            return new VarselbrevManglendeInnbetalingBrevbestilling(this);
        }
    }
}

File: domain/src/main/kotlin/no/nav/melosys/domain/brev/VarselbrevManglendeInnbetalingBrevbestilling.kt
package no.nav.melosys.domain.brev

import java.time.LocalDate
import no.nav.melosys.domain.manglendebetaling.Betalingsstatus

class VarselbrevManglendeInnbetalingBrevbestilling : DokgenBrevbestilling {
    var fakturanummer: String? = null
    var betalingsstatus: Betalingsstatus? = null
    var fullmektigForBetaling: String? = null
    var betalingsfrist: LocalDate? = null
    var erEøsPensjonist: Boolean? = null
    var erEøsLovvalg: Boolean? = null

    constructor() : super()

    constructor(builder: Builder) : super(builder) {
        this.fakturanummer = builder.fakturanummer
        this.betalingsstatus = builder.betalingsstatus
        this.fullmektigForBetaling = builder.fullmektigForBetaling
        this.betalingsfrist = builder.betalingsfrist
        this.erEøsPensjonist = builder.erEøsPensjonist
        this.erEøsLovvalg = builder.erEøsLovvalg
    }

    override fun toBuilder(): Builder = Builder(this)

    class Builder : DokgenBrevbestilling.Builder<Builder> {
        internal var fakturanummer: String? = null
        internal var betalingsstatus: Betalingsstatus? = null
        internal var fullmektigForBetaling: String? = null
        internal var betalingsfrist: LocalDate? = null
        internal var erEøsPensjonist: Boolean? = null
        internal var erEøsLovvalg: Boolean? = null

        constructor()

        constructor(varselbrevManglendeInnbetalingBrevbestilling: VarselbrevManglendeInnbetalingBrevbestilling) : super(varselbrevManglendeInnbetalingBrevbestilling) {
            this.betalingsstatus = varselbrevManglendeInnbetalingBrevbestilling.betalingsstatus
            this.fakturanummer = varselbrevManglendeInnbetalingBrevbestilling.fakturanummer
            this.fullmektigForBetaling = varselbrevManglendeInnbetalingBrevbestilling.fullmektigForBetaling
            this.betalingsfrist = varselbrevManglendeInnbetalingBrevbestilling.betalingsfrist
            this.erEøsPensjonist = varselbrevManglendeInnbetalingBrevbestilling.erEøsPensjonist
            this.erEøsLovvalg = varselbrevManglendeInnbetalingBrevbestilling.erEøsLovvalg
        }

        fun medFakturanummer(fakturanummer: String?) = apply { this.fakturanummer = fakturanummer }

        fun medBetalingsstatus(betalingsstatus: Betalingsstatus?) = apply { this.betalingsstatus = betalingsstatus }

        fun medFullmektigForBetaling(fullmektigForBetaling: String?) = apply { this.fullmektigForBetaling = fullmektigForBetaling }

        fun medErEøsPensjonist(erEøsPensjonist: Boolean) = apply { this.erEøsPensjonist = erEøsPensjonist }

        fun medErEøsLovvalg(erEøsLovvalg: Boolean) = apply { this.erEøsLovvalg = erEøsLovvalg }

        fun medBetalingsfrist(betalingsfrist: LocalDate?) = apply { this.betalingsfrist = betalingsfrist }

        override fun build(): VarselbrevManglendeInnbetalingBrevbestilling = VarselbrevManglendeInnbetalingBrevbestilling(this)
    }
}
